File: bintree_java/src/binarytree/traversal/NonRecurTraversal.java
package binarytree.traversal;


import binarytree.BinNode;
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Stack;



public class NonRecurTraversal< TKey extends Comparable<? super TKey>,
                                TNode extends BinNode<TKey, TNode>
                              >
             extends BaseTraversal<TKey, TNode>
{


    @Override
    public List<TKey> traverse(TNode root, OrderTraversal order) {
        var resPath = new ArrayList<TKey>();

        if (null == root)
            return resPath;

        switch (order) {
        case PRE:
            traversePre(root, resPath);
            break;

        case IN:
            traverseIn(root, resPath);
            break;

        case POST:
            traversePost(root, resPath);
            break;

        default:
            break;
        }

        return resPath;
    }



    protected void traversePre(TNode node, List<TKey> resPath) {
        resPath.clear();
        var stack = new Stack<TNode>();

        stack.push(node);

        while (false == stack.empty()) {
            node = stack.pop();
            resPath.add(node.key);

            if (null != node.right)
                stack.push(node.right);

            if (null != node.left)
                stack.push(node.left);
        }
    }



    protected void traverseIn(TNode node, List<TKey> resPath) {
        resPath.clear();
        var stack = new Stack<TNode>();

        while (true) {
            while (null != node) {
                stack.add(node);
                node = node.left;
            }

            if (stack.empty()) {
                break;
            }

            TNode pickedNode = stack.pop();
            resPath.add(pickedNode.key);

            node = pickedNode.right;
        }
    }



    protected void traversePost(TNode node, List<TKey> resPath) {
        resPath.clear();
        var stack = new Stack<TNode>();

        stack.push(node);

        while (false == stack.empty()) {
            node = stack.pop();
            resPath.add(node.key);

            if (null != node.left)
                stack.push(node.left);

            if (null != node.right)
                stack.push(node.right);
        }

        Collections.reverse(resPath);
    }
}

File: bintree_csharp/binarytree/traversal/NonRecurTraversal.cs
using System;
using System.Collections.Generic;


namespace my.binarytree.traversal
{
    class NonRecurTraversal<TKey, TNode> : BaseTraversal<TKey, TNode>
        where TKey : IComparable where TNode : BinNode<TKey, TNode>
    {
        public override List<TKey> Traverse(TNode root, OrderTraversal order)
        {
            var resPath = new List<TKey>();

            if (root is null)
                return resPath;

            switch (order)
            {
                case OrderTraversal.PRE:
                    TraversePre(root, ref resPath);
                    break;

                case OrderTraversal.IN:
                    TraverseIn(root, ref resPath);
                    break;

                case OrderTraversal.POST:
                    TraversePost(root, ref resPath);
                    break;

                default:
                    break;
            }

            return resPath;
        }



        protected void TraversePre(TNode node, ref List<TKey> resPath)
        {
            resPath.Clear();
            var stack = new Stack<TNode>();

            stack.Push(node);

            while (stack.Count > 0)
            {
                node = stack.Pop();
                resPath.Add(node.Key);

                if (node.Right is not null)
                    stack.Push(node.Right);

                if (node.Left is not null)
                    stack.Push(node.Left);
            }
        }



        protected void TraversePost(TNode node, ref List<TKey> resPath)
        {
            resPath.Clear();
            var stack = new Stack<TNode>();

            stack.Push(node);

            while (stack.Count > 0)
            {
                node = stack.Pop();
                resPath.Add(node.Key);

                if (node.Left is not null)
                    stack.Push(node.Left);

                if (node.Right is not null)
                    stack.Push(node.Right);
            }

            resPath.Reverse();
        }



        protected void TraverseIn(TNode node, ref List<TKey> resPath)
        {
            resPath.Clear();
            var stack = new Stack<TNode>();

            while (true)
            {
                while (node is not null)
                {
                    stack.Push(node);
                    node = node.Left;
                }

                if (0 == stack.Count)
                    break;

                var pickedNode = stack.Pop();
                resPath.Add(pickedNode.Key);

                node = pickedNode.Right;
            }
        }
    }
}
